### PR TITLE
[WIP] BlockOut for Jasper

### PR DESCRIPTION
**Pull request by AlexGrinch**
BlockOut is a regularization technique which drops whole blocks from the graph utilizing the residual connections between them. Has 2 parameters:
-- drop_block_prob: float, probability of dropping blocks during training
-- drop_block_index: int, index of block to drop on inference

### CODE REVIEW

File: open_seq2seq/encoders/tdnn_encoder.py [hunk 1]
```diff
@@ -28,6 +28,8 @@ def get_optional_params():
         'bn_momentum': float,
         'bn_epsilon': float,
         'use_conv_mask': bool,
+        'drop_block_prob': float,
```
Comment (blisc) [inline open_seq2seq/encoders/tdnn_encoder.py]: We should document how to use these settings in the init.
I should also document what use_conv_mask is.

File: open_seq2seq/encoders/tdnn_encoder.py [hunk 2]
```diff
@@ -195,6 +201,8 @@ def _encode(self, input_dict):
               regularizer=regularizer,
               training=training,
               data_format=data_format,
+              drop_block_prob=drop_block_prob,
+              drop_block=(drop_block_index == idx_convnet),
```
Comment (blisc) [inline open_seq2seq/encoders/tdnn_encoder.py]: drop_block_index which by default is -1 indicates that no blocks are ever dropped.
Is this the intended behaviour?

File: scripts/run_all_tests.sh [hunk 3]
```diff
@@ -64,3 +64,12 @@ scripts/multi-bleu.perl toy_text_data/test/target.txt < log_text2text_small/TT.o
...
 
+python run.py --config_file=open_seq2seq/test_utils/test_speech_configs/jasper_res_blockout_test_config.py \
+  --mode=train_eval --encoder_params/drop_block_prob=0.98
+python run.py --config_file=open_seq2seq/test_utils/test_speech_configs/jasper_res_blockout_test_config.py \
+  --mode=eval --encoder_params/drop_block_index=0
+python run.py --config_file=open_seq2seq/test_utils/test_speech_configs/jasper_res_blockout_test_config.py \
+  --mode=eval --encoder_params/drop_block_index=1
+python run.py --config_file=open_seq2seq/test_utils/test_speech_configs/jasper_res_blockout_test_config.py \
+  --mode=eval --encoder_params/drop_block_index=2
+rm -rf tmp_log_folder/
```
Comment (blisc) [inline scripts/run_all_tests.sh]: @borisgin In our documentation, we never say to run this file for unit testing. We only say to run `python -m unittest discover -s open_seq2seq -p '*_test.py'` (which means that the new jasper blockout tests will not run).
Is our documentation out of date or is this script out of date?

File: open_seq2seq/encoders/tdnn_encoder.py (hunk 2 above)
Comment (AlexGrinch) [reply open_seq2seq/encoders/tdnn_encoder.py]: drop_block_index corresponds to dropping blocks at inference only. During training the mechanism of dropping blocks is controlled entirely by parameter drop_block_prob.

File: open_seq2seq/encoders/tdnn_encoder.py (hunk 1 above)
Comment (AlexGrinch) [reply open_seq2seq/encoders/tdnn_encoder.py]: drop_block_prob controls the probability of dropping blocks on training, drop_block_index defines which block to drop on inference. Thus, the default parameters which will correspond to turned off BlockOut are drop_block_prob=0.0, drop_block_index=-1

**Pull request merged**